pkg/controller/cloudinit/controller.go

```diff
@@ -0,0 +1,403 @@
+package cloudinit
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	ctlnodev1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cloudinitv1 "github.com/harvester/node-manager/pkg/apis/node.harvesterhci.io/v1beta1"
+	"github.com/harvester/node-manager/pkg/cloudinit"
+	ctrlv1 "github.com/harvester/node-manager/pkg/generated/controllers/node.harvesterhci.io/v1beta1"
+)
+
+const (
+	handlerName = "harvester-node-cloud-init-controller"
+
+	// CloudInit objects are Cluster-scoped, but Kubernetes really wants
+	// a Namespace associated with the Event objects. It rejects an Event
+	// object with a Namespace of `harvester-system` since the CloudInit
+	// object is Cluster-scoped and hence not in `harvester-system` or any
+	// Namespace. Kubernetes is fine with `default` though.
+	eventNamespace = "default"
+
+	eventActionReconcile = "ReconcileContents"
+	eventReasonReconcile = "CloudInitFileModified"
+	eventActionRemove    = "RemoveFile"
+	eventReasonRemove    = "CloudInitNotApplicable"
+)
+
+type controller struct {
+	nodeName   string
+	cloudinits ctrlv1.CloudInitClient
+	events     ctlnodev1.EventClient
+	nodeCache  ctlnodev1.NodeCache
+}
+
+func Register(ctx context.Context, nodeName string, cloudinits ctrlv1.CloudInitController, nodeCache ctlnodev1.NodeCache, events ctlnodev1.EventClient) {
+	ctl := &controller{
+		nodeName:   nodeName,
+		cloudinits: cloudinits,
+		events:     events,
+		nodeCache:  nodeCache,
+	}
+
+	cloudinits.OnChange(ctx, handlerName, ctl.OnCloudInitChange)
+	cloudinits.OnRemove(ctx, handlerName, ctl.OnCloudInitRemove)
+}
+
+func (c *controller) OnCloudInitChange(_ string, cloudInitObj *cloudinitv1.CloudInit) (*cloudinitv1.CloudInit, error) {
+	if cloudInitObj == nil || cloudInitObj.DeletionTimestamp != nil {
+		return cloudInitObj, nil
+	}
+
+	cloudInitCopy := cloudInitObj.DeepCopy()
+
+	if cloudInitCopy.Annotations == nil {
+		cloudInitCopy.Annotations = make(map[string]string)
+	}
+
+	checksum, err := cloudinit.Measure(strings.NewReader(cloudInitCopy.Spec.Contents))
+	_ = err
+
+	checksumString := fmt.Sprintf("%x", checksum)
+	persistedChecksum := strings.ToLower(cloudInitCopy.Annotations[cloudinit.AnnotationHash])
+
+	if checksumString != persistedChecksum {
+		cloudInitCopy.Annotations[cloudinit.AnnotationHash] = checksumString
+		return c.cloudinits.Update(cloudInitCopy)
+	}
+
+	node, err := c.nodeCache.Get(c.nodeName)
+	if err != nil {
+		return cloudInitCopy, err
+	}
+
+	if cloudInitCopy.Spec.Paused {
```

Comment:
I got confused here.

When we paused the cloudinit CR, we skip do anything. Is it what you want?
If yes, could we check it more early? Also, we can return nil.

If we try to update and return the update result, that will trigger the OnChange again.

Reply:
> When we paused the cloudinit CR, we skip do anything. Is it what you want?

Right, but more specifically I want the controller to not modify any files on disk,

>  If yes, could we check it more early? Also, we can return nil.

but I still want it to make sure the checksum annotation is correct on the resource and to make sure the status is up-to-date, so we can't return earlier as we need the checksum reconciled as well as the status.

Reply:
Monitoring the checksum seems to make sense.
But, IIRC, any change for the cloud-init CR will trigger the OnChange again. So it will check the checksum at that time.

Keep return c.updateStatus here might loop the OnChange.

cc @ibrokethecloud if I got something wrong with that mechanism

Reply:
the status would not have changed and that will return no status updates and it will reconcile successfully.

Reply:
Got it; thanks for the explanation.